File: Go/XApp/shared_test.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XApp

import (
	"sync"
	"testing"
	"time"

	"github.com/frameworkease/Universal.X.Utility/Go/XLog"
	"github.com/stretchr/testify/assert"
)

type MyApplication struct {
	started bool
	stopped bool
	mu      sync.Mutex
}

func (application *MyApplication) Awake() bool {
	return true
}

func (application *MyApplication) Start() {
	application.mu.Lock()
	defer application.mu.Unlock()
	application.started = true
	XLog.Notice("MyApplication started.")
}

func (application *MyApplication) Stop(wg *sync.WaitGroup) {
	application.mu.Lock()
	defer application.mu.Unlock()
	wg.Add(1)       // 先增加计数
	defer wg.Done() // 确保在函数返回时减少计数
	application.stopped = true
	XLog.Notice("MyApplication stopped.")
}

func TestLifecycle(t *testing.T) {
	var onAwakeEventInvoked bool
	var onStartEventInvoked bool
	var onStopEventInvoked bool
	Event.Register(int(EventOnAwake), func(args ...any) {
		onAwakeEventInvoked = true
	})
	Event.Register(int(EventOnStart), func(args ...any) {
		onStartEventInvoked = true
	})
	Event.Register(int(EventOnStop), func(args ...any) {
		onStopEventInvoked = true
		if len(args) > 0 {
			if wg, ok := args[0].(*sync.WaitGroup); ok {
				wg.Add(1)
				go func() {
					time.Sleep(100 * time.Millisecond)
					wg.Done()
				}()
			}
		}
	})

	application := &MyApplication{}
	done := make(chan struct{})

	go func() {
		Run(application)
		close(done)
	}()

	// 等待应用程序启动
	time.Sleep(500 * time.Millisecond)

	application.mu.Lock()
	assert.True(t, application.started, "应用程序应当启动。")
	application.mu.Unlock()
	assert.True(t, onAwakeEventInvoked, "应用程序初始化事件应当被调用。")
	assert.True(t, onStartEventInvoked, "应用程序启动事件应当被调用。")

	// 触发退出
	Quit()

	// 等待应用程序完全停止
	select {
	case <-done:
		// 应用程序已经停止
	case <-time.After(time.Second):
		t.Fatal("等待应用程序停止超时。")
	}

	application.mu.Lock()
	assert.True(t, application.stopped, "应用程序应当停止。")
	application.mu.Unlock()
	assert.True(t, onStopEventInvoked, "应用程序退出事件应当被调用。")
}

File: CSharp/tests~/Assets/Tests/Runtime/Scripts/XApp/shared_test.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System.Threading;
using System.Threading.Tasks;
using FrameworkEase.Universal.X.Utility;
using NUnit.Framework;

public class TestXApp
{
    internal class MyApplication : XApp.IBase
    {
        internal bool awaked;
        internal bool started;
        internal bool stopped;

        public bool Awake()
        {
            awaked = true;
            return true;
        }

        public void Start() { started = true; }

        public void Stop(CountdownEvent counter)
        {
            stopped = true; // 设置标志
            counter.AddCount(); // 增加计数
            Task.Run(async () => // 异步停止
            {
                await Task.Delay(100);
                counter.Signal(); // 完成计数
            });
        }
    }

    [Test]
    public async Task Lifecycle()
    {
        var onAwakeEventInvoked = false;
        var onStartEventInvoked = false;
        var onStopEventInvoked = false;
        XApp.Event.Register(XApp.Events.OnAwake, () => { onAwakeEventInvoked = true; });
        XApp.Event.Register(XApp.Events.OnStart, () => { onStartEventInvoked = true; });
        XApp.Event.Register<CountdownEvent>(XApp.Events.OnStop, (counter) =>
        {
            onStopEventInvoked = true; // 设置标志
            counter.AddCount(); // 增加计数
            Task.Run(async () => // 异步停止
            {
                await Task.Delay(100);
                counter.Signal(); // 完成计数
            });
        });

        var myApplication = new MyApplication();
        _ = XApp.Run(myApplication); // 启动应用
        _ = XApp.Run(myApplication); // 重复启动
        await Task.Delay(500); // 等待启动

        Assert.That(XApp.Instance<MyApplication>(), Is.EqualTo(myApplication), "应用程序实例应当相等。");
        Assert.That(myApplication.awaked, Is.True, "应用程序应当初始化。");
        Assert.That(onAwakeEventInvoked, Is.True, "应用程序初始化事件应当被调用。");
        Assert.That(myApplication.started, Is.True, "应用程序应当已启动。");
        Assert.That(onStartEventInvoked, Is.True, "应用程序启动事件应当被调用。");

        XApp.Quit(); // 退出应用
        XApp.Quit(); // 重复退出
        await Task.Delay(500); // 等待退出
        Assert.That(myApplication.stopped, Is.True, "应用程序应当已退出。");
        Assert.That(onStopEventInvoked, Is.True, "应用程序退出事件应当被调用。");
    }
}
